### PR TITLE
[JUJU-1122] Set FORCE-VERSION for uploaded tools;

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR allows juju to be able to upgrade to a local built snap;

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
$ sudo snap install juju --channel 2.9/stable --classic
juju (2.9/stable) 2.9.29 from Canonical✓ installed

$ juju version
2.9.29-ubuntu-amd64

$ which juju
/snap/bin/juju

$ juju bootstrap lxd k1
Creating Juju controller "k1" on lxd/localhost
Looking for packaged Juju agent version 2.9.29 for amd64
Located Juju agent version 2.9.29-ubuntu-amd64 at https://streams.canonical.com/juju/tools/agent/2.9.29/juju-2.9.29-linux-amd64.tgz
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...
 - juju-381a10-0 (arch=amd64)
Installing Juju agent on bootstrap instance
Fetching Juju Dashboard 0.8.1
Waiting for address
Attempting to connect to 10.242.98.130:22
Connected to 10.242.98.130
Running machine configuration script...
Bootstrap agent now started
Contacting Juju controller at 10.242.98.130 to verify accessibility...

Bootstrap complete, controller "k1" is now available
Controller machines are in the "controller" model
Initial model "default" added

$ sudo snap remove juju
juju removed

$ sudo snap install juju_2.9.31_amd64.snap --dangerous --classic
juju 2.9.31 installed

$ juju version
2.9.31-ubuntu-amd64

$ which juju
/snap/bin/juju

## => 1.  upgrade to local snap

$ juju upgrade-controller
no prepackaged agent binaries available, using local agent binary 2.9.31.1
best version:
    2.9.31.1
started upgrade to 2.9.31.1

## => 2.  upgrade to local built

$ $GOPATH/bin/juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 2.9.31.2 (built from source)
best version:
    2.9.31.2
started upgrade to 2.9.31.2

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1812191

### CODE REVIEW

File: cmd/juju/commands/upgrademodel.go [hunk 1]
```diff
@@ -991,27 +999,26 @@ func (context *upgradeContext) maybeChoosePackagedAgent() (err error) {
...
-			}
+			return nil
 		}
-	} else {
-		// If not completely specified already, pick a single tools version.
-		filter := coretools.Filter{Number: context.chosen}
-		if context.packagedAgents, err = context.packagedAgents.Match(filter); err != nil {
-			return errors.Wrap(err, errors.New("no matching agent versions available"))
+		newestCurrent, found := context.packagedAgents.NewestCompatible(context.agent)
+		if found && newestCurrent.Compare(context.agent) > 0 {
```
Comment (wallyworld) [inline cmd/juju/commands/upgrademodel.go:1005]: do we need `newestCurrent.Compare(context.agent) > 0`?
Doesn't `NewestCompatible` give us a newestCurrent that is always > context.agent?

File: environs/tools/build.go [hunk 2]
```diff
@@ -217,6 +217,9 @@ func copyExistingJujus(dir string) error {
 			return errors.Trace(err)
 		}
 	}
+	if skipCopyVersionFile {
```
Comment (wallyworld) [inline environs/tools/build.go:220]: I can't see a test for this?

**Review by wallyworld — APPROVED** [review]

File: cmd/juju/commands/upgrademodel.go (hunk 1 above)
Comment (ycliuhw) [reply cmd/juju/commands/upgrademodel.go:1005]: It could return a version equal to `context.agent`.

File: environs/tools/build.go (hunk 2 above)
Comment (ycliuhw) [reply environs/tools/build.go:220]: right, added one more test to cover this

**Comment by manadart** [comment]
Can we rebase instead of merging to update in-flight branches? It obfuscates the change history.

**Pull request merged**